core/src/main/java/org/infinispan/executors/LimitedExecutor.java

```diff
@@ -0,0 +1,156 @@
+package org.infinispan.executors;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.jboss.logging.NDC;
+
+/**
+ * Executes tasks in the given executor, but never has more than {@code maxConcurrentTasks} tasks running at the same
+ * time.
+ *
+ * The number of concurrent tasks can be tweaked at any time with {@link #addPermit()} and {@link #removePermit()}.
+ *
+ * @author Dan Berindei
+ * @since 9.0
+ */
+public class LimitedExecutor implements Executor {
```

Comment:
Because I still want to limit the number of tasks that run concurrently. A `LimitedExecutor(1)` allows me to submit 2 tasks and runs only 1 at a time, but without creating a new thread for each `LimitedExecutor` instance. `executeAsync` is also something you don't get in a regular `Executor` :)